### PR TITLE
doc/stdenv: rewrite manual build procedure to be closer to an auto-build

### PR DESCRIPTION
**Pull request by lf-**
This is based on
<https://jade.fyi/blog/building-nix-derivations-manually/> plus some more original research.

The previous version of this section did not work for your choice of simple Haskell package, e.g. haskellPackages.hscolour, due to things like `compileBuildDriverPhase` and other custom phases that it does not address at all.

It seems more correct to use genericBuild in development to harmonize it with what is actually done.

I feel a little bit like I am committing a sin by suggesting using the experimental CLI in the manual (afaict I am the first to do this), but I have given the old version of the command, and there are justifiable reasons to do it:
* The noted limitations with env-vars are fixed. The one with the non-empty temp directory was one I ran into myself and oh boy was that not fun to debug.
* Additionally the outputs are set *before* sourcing `setup.sh`: there is an issue with nix-shell where the original version of `$out` winds up in `NIX_LDFLAGS` due to _addRpathPrefix, which means that resulting executables may not run properly.

It is sad that `nix develop` propagates a wrong value of `SHELL` to builders, though. It is equally sad that `nix-shell` is essentially abandoned upstream, with undocumented and not insignificant differences from `nix develop`.

For the exact script differences:
https://github.com/NixOS/nix/blob/17e6b85d05b3d32df244b1d4e89aa41fd8bdcae8/src/nix-build/nix-build.cc#L516-L551
https://github.com/NixOS/nix/blob/db026103b18fb8b5a719594502edd0f89eb9c268/src/nix/get-env.sh

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by fricklerhandwerk — CHANGES_REQUESTED** [review]
@lf- thanks a lot, this is a clear improvement.

I have a stylistic change request: Please use plain text to explain commands instead of adding comments. It's significantly easier to read for people who are not fluent with the system yet (the distinction between regular "speech" and code seems to do things to our brains). It would be more work for you now, but since this is a central piece of documentation it will benefit countless beginners as time passes.

Please tell me if you don't have time to address the review comments, then I'll put it on my stack to take over once I get a couple of quiet minutes at the keyboard.

Concerning the maintenance state of `nix-shell` and `nix-develop` I have a possibly controversial opinion: development on both stalled, because there are too many tricky issues, and those all arise because Nix is trying to wrap a Nixpkgs feature. Hooking into stdenv builds should be done by stdenv facilities exclusively, and Nix should  only run a special executable that does this and is provided by the derivation in question. In other words, Nix should be fully oblivious to the existence of stdenv. Now this is not going to change anytime soon, because for that, Nix maintainers would have to let go of that feature, somewhat gracefully in fact, and someone would have to implement it in Nixpkgs.

What I'm trying to say: I'd prefer to stick to what works now and will surely work unchanged for a long time.
We can't count on any particular change to happen because it requires effort that may or may not be affordable at any given time - including fixing documentation of moving targets when it gets stale or irrelevant – but we can count on existing things to continue working, because that's kind of what Nix promises.

If you strongly disagree we can escalate this to the documentation team.

File: doc/stdenv/stdenv.chapter.md
Comment (lf-) [reply doc/stdenv/stdenv.chapter.md]: added

**Comment by lf-** [comment]
Alright, I think I have addressed your feedback. Thanks for the thorough review! Feel free to directly commit any further stylistic changes you need on this before merging, if any arise.

File: doc/stdenv/stdenv.chapter.md
Comment (fricklerhandwerk) [inline doc/stdenv/stdenv.chapter.md:128]:
```suggestion
Re-run a single phase as many times as necessary to examine the failure like so:
```

File: doc/stdenv/stdenv.chapter.md
Comment (fricklerhandwerk) [inline doc/stdenv/stdenv.chapter.md:156]:
```suggestion
If the build fails differently inside the shell than in the sandbox, consider using [`breakpointHook`](#breakpointhook) and invoking `nix-build` instead.
```

File: doc/stdenv/stdenv.chapter.md
Comment (fricklerhandwerk) [inline doc/stdenv/stdenv.chapter.md:157]:
```suggestion
The [`--keep-failed`](https://nixos.org/manual/nix/unstable/command-ref/conf-file#opt--keep-failed) option for `nix-build` may also be useful to examine the build directory of a failed build.
```

**Review by fricklerhandwerk — APPROVED** [review]

**Pull request merged**
